tao_compiler/mlir/xla/ral/context/common_context_impl_quantization.cc

```diff
@@ -726,6 +725,102 @@ void ral_qgemm_onednn_s8_s8_s8_per_channel(
+    ExecutionContext* ctx, opaque_t /*stream_handle*/,
+    MemRefType<int8_t, 2> input, MemRefType<int8_t, 2> weight,
+    MemRefType<float, 1> bias, MemRefType<float, 0> inputScales,
+    MemRefType<int32_t, 0> inputZeroPoints, MemRefType<float, 1> weightScales,
+    MemRefType<int32_t, 1> weightZeroPoints, MemRefType<float, 0> resultScales,
+    MemRefType<int32_t, 0> resultZeroPoints, void* customAttrs) {
+  CpuTimer timer("ral_pdll_qgemm_onednn_s8_s8_s8_per_channel");
+  int64_t resultSizes[2] = {0, 0};
+  if (isEmptyMemref(input) || isEmptyMemref(weight)) {
+    TAO_VLOG(1)
+        << "ral_pdll_qgemm_onednn_s8_s8_s8_per_channel: early return for "
+           "empty tensor";
+    return assignMemRef<int8_t, 2>(nullptr, resultSizes);
+  }
+  if (TAO_VLOG_IS_ON(1)) {
+    for (int i = 0; i < Size(input); ++i) {
+      TAO_VLOG(0) << "input[" << i
+                  << "] = " << static_cast<int32_t>(input.data[i]);
+    }
+    for (int i = 0; i < Size(weight); ++i) {
+      TAO_VLOG(0) << "weight[" << i
+                  << "] = " << static_cast<int32_t>(weight.data[i]);
+    }
+    for (int i = 0; i < Size(bias); ++i) {
+      TAO_VLOG(0) << "bias[" << i << "] = " << static_cast<float>(bias.data[i]);
+    }
+  }
+  auto attr = getOrParsePDLAttr(ctx, customAttrs,
+                                "ral_pdll_qgemm_onednn_s8_s8_s8_per_channel");
+  if (!attr) {
+    ctx->signalError(Context::FAILURE, "fail to parse custom_attrs\n");
+  }
+  auto& dictAttr = attr->as<DictPDLAttr>();
+  bool tp_a = dictAttr.get("transpose_a").as<BoolPDLAttr>().getValue();
+  bool tp_b = dictAttr.get("transpose_b").as<BoolPDLAttr>().getValue();
+  if (tp_a || tp_b) {
+    ctx->signalError(Context::FAILURE,
+                     "not supported ral_qgemm with transpose");
+  }
+  int64_t m = tp_a ? input.sizes[1] : input.sizes[0];
+  int64_t k = tp_a ? input.sizes[0] : input.sizes[1];
+  if (k != (tp_b ? weight.sizes[1] : weight.sizes[0])) {
+    ctx->signalError(Context::FAILURE,
+                     "mismatch contraction dim for quantization gemm");
+  }
+
+  int64_t n = (tp_b ? weight.sizes[0] : weight.sizes[1]);
+  resultSizes[0] = m;
+  resultSizes[1] = n;
+  auto driver = ctx->getDriver<cpu::CPUDriver>(cpu::CPUDriver::name());
+  auto data = static_cast<int8_t*>(driver->alloc(ctx, m * n * sizeof(int8_t)));
+  auto result = assignMemRef<int8_t, 2>(data, resultSizes);
+
+  ideep::tensor input_t{dims{m, k}, ideep::data_type::s8,
+                        tp_a ? format_tag::ba : format_tag::ab, input.data};
+  ideep::tensor weight_t{dims{k, n}, ideep::data_type::s8,
+                         tp_b ? format_tag::ba : format_tag::ab, weight.data};
+  ideep::tensor bias_t{dims{1, n}, ideep::data_type::f32, format_tag::ab,
+                       bias.data};
+  ideep::tensor output_t{dims{m, n}, ideep::data_type::s8, format_tag::ab,
+                         result.data};
+
+  std::vector<float> input_scales({inputScales.data[0]});
```

Comment:
If not online quantization, scale/zero_points are all constants. One possible optimization is to make these settings as attributes instead of inputs of the kernel. And then `input_scales ` and all other similar settings can be cached to reduce overhead. This can also help to reduce const op dispatch overhead in DISC. Current implementation is ok for PR and we may choose to do it in the future.

Reply:
Agree, I will optimize the implementation in the future pr.